### ISSUE TITLE
Understanding Data-Parallel Optimizations

### ISSUE THREAD

**Issue by joelsimonoff** [aws/aws-neuron-sdk#243]
This is a mix of a feature request and a question so please assign whatever issue tag you feel makes the most sense or feel free to close it if its too far outside the scope of Github issues.

### Motivation

Optimizing only the compiled batch size seems to leave a lot of performance on the table. Our initial attempts at optimized data-parallel give us an additional 27% speedup over just optimizing the compiled batch size. We are fairly convinced there are lots more performance boosts to be had; however, we are having difficulties identifying how to get that performance.

I'll walk through what we've done so far and then ask targeted questions that are blocking us and make a recommendation to consider updating the tutorials to make this easier for future teams.

### Reference

- Model:  RoBERTa for sequence classification with a max sequence length of 256.
- Framework: PyTorch
- simple data-parallel: we are referring to the class in the Neuron documentation https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/pytorch-neuron/tutorials/getting_started.html#pytorch-getting-started
- Machine: inf1.6xlarge
- Optimal Batch Size: 4 – found via experimentation
- All data is stored in RAM

Additionally, our workload is to run as many inferences as possible in one hour. Therefore we care about high throughput but don't care about latency.

### Optimizing Batch Size
**Optimizing batch size gives us 42% speedup (when testing on 16 neuron cores).**

Using a single Neuron Core, we've found that optimizing the batch size can give us a performance boost of roughly 16.5% speedup 11.2 ms per inference to 9.35 ms per inference (each batch of 4 took 37.4 ms). When using simple data-parallel with 16 neuron cores we see a 42% speedup 1.64 ms/inference to 0.95 ms/inference with optimal batch size.

### Optimizing Data Parallel

_This section is based on batch size 4._

Using the simple data-parallel class our total run time for 16 Neuron cores was 60.6ms vs 37.4ms on 1 core. We felt that this task should be almost perfectly parallelizable therefore we would expect 16 Neuron cores to have very close to the same runtime as a single core. Our best guess is that the simple data-parallel function had some sort of runtime fraction not devoted to inteferencing such that when we looped over simple data-parallel (each batch is one iteration of the loop) then our average batch time worked out to 60.6ms. Since the models are only initialized when the class is first created and not during each iteration it couldn't be a model load issue. We then identified the most likely culprit being that some neuron cores inference faster than others and if we call simple data-parallel every batch then we need to wait for every thread to finish before starting the next batch – this is just our hypothesis we haven't proven this is actually the cause.

Given that the page mentions: `This is intended as a good starting implementation - but you may want to vary it depending on your application` and we think we identified the bottleneck, therefore we decided to try optimizing it.

#### Optimization 1

The first optimization we tried was adding dynamic batching to the `SimpleWrapper` class. 

Initial psuedo-code:

```
 for batch in batches:
    simple_data_parallel_model(batch)
```

New psuedo-code:

```
new_data_parallel(batches)
```

The modification results in dividing the batches evenly between all neuron cores. Each neuron core has its own thread and in its thread, it loops through each share of the batches. Therefore if we have 16 neuron cores and 16000 batches then we will have 16 threads (one per core) and each thread inferences 1k batches on its respective neuron core.

**Results**

Our goal was to get as close to 37.4ms per batch as possible.

First, we tried with 4 neuron cores thus we had a batch size of 16 (batch size of 4 per neuron core)

| Nrof Batches  | Time Per Batch (ms) |
| ------------- | ------------- |
| 1 | 50.7  |
| 4 | 46.25  |
| 16 | 43.6 |
| 32 | 42.5 |
| 256 | 41.2 |
| 1024 | 41.2 |

Next, we tried with 16 neuron cores thus our batch size is now 64 (4 per core and 16 cores)

| Nrof Batches  | Time Per Batch (ms) |
| ------------- | ------------- |
| 1 | 60.6  |
| 4 | 51.5  |
| 16 | 46.3 |
| 32 | 45.0 |
| 256 | 43.36 |
| 1024 | 44.6 |

Our observation is that the cold start problem gets worse with more cores and after 256 batches using this optimization the effects of the cold start seems to be completely mitigated. However, we still felt like we should be able to do much better because the 16 core best average batch time is 43.36ms and we think we should be able to get closer to 37.4ms.

#### Optimization 2

The second optimization we tried seems to be more in line with what I've seen in some Inferentia tutorials. We went back to one batch per threads and just made the number of threads equal to the number of batches. We feel this may not scale because you might need to create too many threads when given a large workload.

New psuedo-code:
```
models = [SimpleWrapper(torch.jit.load(model_file)) for i in range(num_neuron_cores)]
for i, batch in batches:
    create_thread(models[i % num_neuron_cores], batch)
```

**Results**

Again we started with 4 neuron cores thus we had a batch size of 16 (batch size of 4 per neuron core)

| Nrof Batches  | Time Per Batch (ms) |
| ------------- | ------------- |
| 4 | 47.5  |
| 1024 | 41.6 |

Next, we moved to 16 neuron cores thus our batch size is now 64 (4 per core and 16 cores)

| Nrof Batches  | Time Per Batch (ms) |
| ------------- | ------------- |
| 4 | 52  |
| 1024 | 45.3 |

Here we saw relatively similar performance; however, optimization 2 always did very slightly worse. We felt this was within the margin of error of testing so we were unable to conclusively determine if this worked better than the first optimization. Given in the real world we'll have tens of millions of batches for each job so this may run into problems with millions of threads.

#### Conclusions From Optimizing Data Parallel

We think that optimization 1 was a better solution than optimization 2. When running Optimization 1 we were able to run 65536 inferences (1024 batches of size 64) in 45.7 seconds (44.6 ms/batch) which is roughly 0.69 ms/inference. Breaking down the speed up:

Performance Vs Compiled BS 1 (not optimal batch size) on 16 neuron cores:
0.69 ms/inference vs 1.64 ms/inference = 58% speedup

Performance Vs Compiled BS 4 (optimal batch size) on 16 neuron cores:
0.69 ms/inference vs 0.95 ms/inference = 27% speedup

**Thus optimizing the compiled batch size and data-parallel gives us a 58% speedup, but only optimizing batch size gives us a measly 42% speedup.** However, we still feel there are some performance improvements left on the table for more optimized data-parallel.

### Questions

Given all of the experimentation results above, I don't think we understand enough about why we are having parallelization issues to arrive at the most optimal solution. Therefore your insight is invaluable to us.

1. Since these workloads are close to perfectly parallelizable and in both simple data-parallel and our optimized data-parallel approaches, we see large differences between 1 core and multi-core inference times, why is this? (i.e.) we are hoping that we can inference on 16 cores at as close as possible to the 1 core inference time.
2. What is the recommended optimal data-parallel approach and why does it work better than the optimizations we suggested?

### Feature Requests

1. Replace the current NeuronSimpleDataParallel class, https://awsdocs-neuron.readthedocs-hosted.com/en/latest/neuron-guide/neuron-frameworks/pytorch-neuron/tutorials/getting_started.html#pytorch-getting-started, with a more optimized version. I think this is beneficial because it doesn't reduce performance on the single batch inference and gives 27% boost on multi-batch runs. This may help future teams get better performance out of the gate.
2. If there exists an optimal approach to data-parallel, it would be awesome if that could get added to the performance tuning section of the docs?

**Comment by mrnikwaws** [comment]
Hi Joel,

Thanks for the thoughtful analysis.  

1. Which RoBERTa model did you use? https://huggingface.co/transformers/model_doc/roberta.html?  If so which version of transformers were you working with?
2. Are you able to share the code you used for compiling and running Roberta on Neuron?  Your pseudo-code is helpful, but if possible it would be better to talk about concrete code

I'll take some time to look into performance for this model - and see if we can make some more recommendations for performance.

**Comment by joelsimonoff** [comment]
- transformers Version: 4.2.2
- torch: 1.7.1
- torch-neuron: 1.7.1.1.1.7.0
- neuron-cc: 1.1.7.0+6d41df113

Model: [RobertaForSequenceClassification](https://huggingface.co/transformers/master/model_doc/roberta.html#transformers.RobertaForSequenceClassification)

Everything lines up exactly with the BERT auto partitioning tutorial so compiled using `model_neuron = torch.neuron.trace(model, example_inputs,  compiler_args=["-O2"])`. The only deviation from the tutorial is the checkpoint. The checkpoint we loaded was trained using Huggginfaces mixed-precision trainer (but I don't think that should affect parallelization performance).

**Comment by aws-joshim** [comment]
Hi Joel, we have observed that data parallel execution using PyTorch python threads doesn't produce linear scaling in performance, especially with 16 cores. We are exploring alternatives to address the threading bottleneck by spawning independent processes to run inference on a single NeuronCore, keeping the resources as independent as possible.
Our initial findings show that replacing the thread pool executor with process pool, and passing static batch to each process can lead to improved performance. We will keep this issue posted with timelines for tutorial updates on data parallel performance improvements.

**Comment by joelsimonoff** [comment]
@aws-joshim thanks for following up! I apologize for not being able to get more testing done, I've been out of the office for the last two weeks and just got back today so perfect timing. 

What speedups are you guys seeing from multi-processing? 

I ran some experiments where I launched each inferencer as a separate process using bash to completely remove any python issues – we pre-segment the input data prior to launching the processes and each program reads a different section of the data. We saw very minimal speedups we went from 43-46 ms per batch using the threading batch data-parallel approach above to 41-43 ms per batch when launching each inferencer as a separate process. However, we noticed that the performance is dependent on the number of PyTorch threads in each process. If you use: `torch.set_num_threads(24)` it takes much longer than `torch.set_num_threads(1)` and `torch.set_num_threads(3)` seems to be about optimal on 16 neuron cores. We did some experiments with different combinations of PyTorch threads and neuron cores.

Here are some interesting results (using model compiled for batch size 4 and multi-processing for data parallel):

| Nrof Neuron Cores | Nrof PyTorch Threads  | Avg Neuron Core % | Avg Inference Time (ms) | 
| ------------- | ------------- | ------------- | ------------- |
| 1 |  1  | 83% | 40.3  |
| 1 | 12 | 90% | 37.6 |
| 1 | 24  | 90% | 37.6 |
| 2 |  1  | 83% | 41.0  |
| 2 | 12 | 88% | 38.7 |
| 2 | 24  | 85% | 39.5 |
| 4 |  1  | 83% | 42.8  |
| 4 | 12 | 83% | 40.9 |
| 4 | 24  | 56% | 60.7 |
| 16 |  1  | 79% | 44.9 |
| 16 | 2 | 83% | 42.6 |
| 16 | 3 | 85% | 42.3 |
| 16 | 12 | 17% | too long |
| 16 | 24  | 8% | too long |

_Notes:_
- **Avg Inference Time is the amount of time for 1 batch to run on each neuron core. For example: it takes an average of 42.3ms for 1 batch to run on 16 neuron cores and each neuron core has an avg utilization of 85%**.
- These values were calculated over many trials. Prior to the trials we ran 1k inferences on boot to eliminate any cold start delays before measuring.
- For 16 neuron cores and 12/24 threads our tester was going taking too long to run so we didn't wait for it.
- We think that the Avg Inference Times are likely only accurate to within +- 0.5 ms.

### Conclusions

We think the results are quite conclusive that the main bottleneck is the operators running on CPU the model only compiles 717 of 722 arithmetic operators to inferentia). Our next set of experiments will be focused on tracking down which specific operations are causing the problem and optimizing the compilation to fix it. As we go forward we are excited to try out some of the graph debugging and profile tools you have built! As another feature request maybe a future inferentia instance with more CPUs to neuron cores might via useful.

Additionally, we found that there is some benefit to processing but its less significant than the CPU operations. We ran it multiple times, and it's consistent so it's imho on the edge of statistically significant results. If you got similar speedups from processing then that can validate that the threading bottleneck exists but it's minimal. Maybe we would see better results from multi-processing in Python directly instead of using bash to spawn multiple processes.

@aws-joshim do you agree with these findings, do they match your team's internal work?

**Comment by aws-joshim** [comment]
Hi Joel,
We profiled and ran benchmark tests for the Roberta Hugging Face sequence classification model. We were able to get a 37.7ms average latency with a producer-consumer implementation with shared input and output queues feeding the NeuronCores.
 
We plan to update the Neuron PyTorch tutorials with the optimized data parallel implementation next. We can share the sample code offline on  aws-neuron-support@amazon.com if you would like to look at our sample code before the tutorials are updated
